### PR TITLE
SHOC grid property tests. 

### PR DESCRIPTION
**Pull request by pressel**
This WIP PR includes first property tests of SHOC's shoc_grid F90 subroutine. At present this PR should be viewed as an example of how to implement property tests of the F90 source code from C++. For now the property tests make simplifying assumptions like fixed grid spacing and uniform density that will likely be eliminated in the future. I am opening this WIP PR for early comment and review prior to finishing code clean-up.

### CODE REVIEW

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      }
+    }
+  }
+  REQUIRE(nerror == 0);
+  
+  //Now check density 
+  for(Int s=0; s < shcol; ++s){
+    for(Int n=0; n < SDS.nlev; ++n){
+      const auto offset = n + s*SDS.nlev;
+      if(SDS.rho_zt[offset] != density){
```
Comment (PeterCaldwell) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:123]: Is it reasonable to expect exact equality here?

Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:123]: I would guess probably not in general, but in this instance with uniform dz and constant density it works.

Comment (PeterCaldwell) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:123]: Ok, I'm just wondering whether it won't work on some machines, or if it is computed in single precision or something. It may be fragile as written.

File: components/scream/src/physics/shoc/shoc_functions_f90.hpp [hunk 2]
```diff
@@ -16,9 +17,54 @@
...
+ Int shcol, nlev, nlevi;
+ Real *zt_grid, *zi_grid, *pdel;
+
+ // In/out
+ Real *dz_zt, *dz_zi, *rho_zt;
+
+
+ SHOCGridData(Int shcol_, Int nevl_, Int nlevi_);
+ SHOCGridData(const SHOCGridData& rhs);
+ SHOCGridData& operator=(const SHOCGridData& rhs);
```
Comment (PeterCaldwell) [inline components/scream/src/physics/shoc/shoc_functions_f90.hpp:38]: Could you explain what the 3 lines above do? Are they ways of initializing the SHOCGridData class? I haven't seen such a thing within a struct before...

Comment (PeterCaldwell) [reply components/scream/src/physics/shoc/shoc_functions_f90.hpp:38]: Nevermind. I got it.

**Comment by PeterCaldwell** [comment]
Do we really need to create SHOCGridData objects in 3 different ways? Why not choose one way and just use that?

**Comment by pressel** [comment]
@PeterCaldwell We aren't really creating them in three different ways. The three functions you are referring to are

1. A parametric constructor that takes parameters of dimensions and allocates the appropriate memory. 
2. A copy constructor describes how the structure is to be copied.  This initializes a new object based on an existing object
3. A copy assignment operator which assigns values of one object into another.   

These are all necessary.

**Comment by ambrad** [comment]
You might find it useful to use and extend [shoc_f90.hpp](https://github.com/E3SM-Project/scream/blob/master/components/scream/src/physics/shoc/shoc_f90.hpp) instead of creating a new grid manager, etc.

File: components/cam/src/physics/cam/shoc.F90 [hunk 3]
```diff
@@ -557,7 +557,7 @@ subroutine shoc_grid( &
 
   ! local variables
   integer :: i, k
-
+
```
Comment (jgfouca) [inline components/cam/src/physics/cam/shoc.F90:560]: Avoid introducing trailing whitespace

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp [hunk 4]
```diff
@@ -50,5 +53,64 @@ void shoc_subroutine(SHOCSubroutineData& d) // example wrapper function
   // shoc_subroutine_c(d.in1, d.in2, d.in3, &d.out1, &d.out2, &d.out3);
 }
 
+SHOCGridData::SHOCGridData(Int shcol_, Int nlev_, Int nlevi_):
+shcol(shcol_), nlev(nlev_), nlevi(nlevi_),
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:57]: Strange indenting scheme. Usually class member initialization has 1 level of indentation.

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp [hunk 5]
```diff
@@ -50,5 +53,64 @@ void shoc_subroutine(SHOCSubroutineData& d) // example wrapper function
...
+m_total(shcol_ * nlev_), m_totali(shcol_ * nlevi_), 
+m_data(NUM_ARRAYS * m_total, 0.0), m_datai(NUM_ARRAYS_i * m_totali, 0.0)
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:59]: Remember to use int literals where possible. `0` instead of `0.0`.

File: components/scream/src/physics/shoc/shoc_functions_f90.hpp [hunk 6]
```diff
@@ -16,9 +17,54 @@
...
+ 
+ void init_ptrs();
+ //void transpose();
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.hpp:41]: This line can be deleted.

File: components/scream/src/physics/shoc/shoc_functions_f90.hpp [hunk 7]
```diff
@@ -16,9 +17,54 @@
 namespace scream {
 namespace shoc {
 
-///////////////////////////////////////////////////////////////////////////////
 
+///////////////////////////////////////////////////////////////////////////////
 // Converted subroutine helpers go here.
+struct SHOCGridData{
+static constexpr size_t NUM_ARRAYS = 4;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.hpp:24]: Need indent for these two lines.

File: components/scream/src/physics/shoc/shoc_functions_f90.hpp [hunk 8]
```diff
@@ -16,9 +17,54 @@
...
 
+///////////////////////////////////////////////////////////////////////////////
 // Converted subroutine helpers go here.
+struct SHOCGridData{
+static constexpr size_t NUM_ARRAYS = 4; 
+static constexpr size_t NUM_ARRAYS_i = 2; 
+
+    
+ // Inputs 
+ Int shcol, nlev, nlevi;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.hpp:29]: Every indentation level should be 2 spaces; it looks like just 1 space here.

File: components/scream/src/physics/shoc/shoc_functions_f90.hpp [hunk 9]
```diff
@@ -16,9 +17,54 @@
...
+   util::transpose<D>(zt_grid, d_trans.zt_grid, shcol, nlev);
+   util::transpose<D>(dz_zt, d_trans.dz_zt, shcol, nlev);
+   util::transpose<D>(pdel, d_trans.pdel, shcol, nlev);
+   util::transpose<D>(rho_zt, d_trans.rho_zt, shcol, nlev);
+
+   // Transpose on the zi grid
+   util::transpose<D>(zi_grid, d_trans.zi_grid, shcol, nlevi);
+   util::transpose<D>(dz_zi, d_trans.dz_zi, shcol, nlevi);
+
+   *this = std::move(d_trans);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.hpp:63]: Nice use of the move assignment. In order for this to work for me, I had to explicitly enable the default move assignment operator. Did you not have to?

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace scream {
+namespace shoc {
+namespace unit_test {
+
+TEST_CASE("shoc_grid", "shoc"){
+
+  constexpr Real gravit = scream::physics::Constants<Real>::gravit;
+
+  const Int shcol = 2;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:30]: These can all be constexprs.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+TEST_CASE("shoc_grid", "shoc"){
+
+  constexpr Real gravit = scream::physics::Constants<Real>::gravit;
+
+  const Int shcol = 2; 
+  const Int nlev = 128; 
+  const auto nlevi = nlev + 1;
+  const Real density = 1.0; 
+
+
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:35]: No need for the 2nd line of whitespace here.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    }
+  }
+
+  // Fill in test data on zi_grid
+  for(Int s=0; s < SDS.shcol; ++s){
+    for(Int n=0; n < SDS.nlevi; ++n){
+       const auto nft = SDS.nlevi - 1 - n;
+       const auto offset = n + s*SDS.nlevi;
+
+       SDS.zi_grid[offset] =  dz * nft;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:60]: I think the contents of this loop can be merged into the loop above.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+ int nerror = 0; 
+ for(Int s=0; s < SDS.shcol; ++s){
+   for(Int n=0; n < SDS.nlev-1; ++n){    
+     const auto offset = n + s*SDS.nlev;
+     if (SDS.zt_grid[offset+1] - SDS.zt_grid[offset] > 0.0){
+          nerror++;
+          }
+      }
+  }  
+  REQUIRE(nerror == 0);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:76]: Since we are not using a kokkos kernel here, there's no need for the nerror pattern. You should just do the REQUIRE directly in the loop. This will result in a more-localized error and a better error message for the user.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+ 
+ //Check that zi decreases upward
+ for(Int s=0; s < SDS.shcol; ++s){
+   for(Int n=0; n < SDS.nlevi-1; ++n){
+     const auto offset = n + s*SDS.nlevi;
+     if (SDS.zi_grid[offset+1] - SDS.zi_grid[offset] > 0.0){
+          nerror++;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:83]: This loop looks exactly identical to the one above. It can be remove if so.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+  //Call the fortran implementation
+  shoc_grid(nlev, SDS);
+
+  //First check that dz is correct
+  for(Int s=0; s < shcol; ++s){
+    for(Int n=0; n < SDS.nlev; ++n){
+     const auto offset = n + s*SDS.nlev;
+      if (SDS.dz_zt[offset] != dz){
+        nerror++;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:97]: No need for nerror pattern.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+  REQUIRE(nerror == 0);
+  
+  for(Int s=0; s < shcol; ++s){
+    const auto offset = s*SDS.nlevi;
+    REQUIRE(SDS.dz_zi[offset] == 0);
+    REQUIRE(SDS.dz_zi[offset + SDS.nlevi-1] == dz/2.0);
+  }
+
+  for(Int s=0; s < shcol; ++s){
+    for(Int n=1; n < SDS.nlevi-1; ++n){
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:110]: Did you intent to start from n=1 instead of n=0 here?

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+      const auto offset = n + s*SDS.nlevi;
+      if (SDS.dz_zi[offset] != dz){
+        nerror++;
+      }
+    }
+  }
+  REQUIRE(nerror == 0);
+  
+  //Now check density 
+  for(Int s=0; s < shcol; ++s){
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:120]: A lot of these loops can be merged. They are all looping over the same ranges.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp [hunk 18]
```diff
@@ -0,0 +1,135 @@
+#include "catch2/catch.hpp"
+
+//#include "share/scream_types.hpp"
+#include "ekat/scream_types.hpp"
```
Comment (bogensch) [inline components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:4]: @pressel I've been focusing on this file.  These tests aren't exactly what I had in mind, but i'm actually glad because it gave me a chance to check this out and give myself a c++/testing crash course.  I have modifications that more reflect my vision for this that run/pass.  Did you want me to commit changes to this branch or hold off and make a separate PR once this goes through?

Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:4]: @bogensch I'm glad you are looking at the tests and have been thinking about extending them. The tests that are implemented are very simple tests to test that the bridge function is working (hence the WIP PR). I do have your specific tests coded up in separate local dev branch and my plan was to go back and add the specific tests you mentioned once we have agreed that the design patterns used are those that we will be sticking with.  That said, if you would like to contribute your tests that is fine with me.  Since @jgfouca has signed off on the design I think we are now at that point.  

Let me finish responding to @jgfouca's review which requires some substantial modifications to the this file, and then we can decide if committing to the branch or a separate PR is best.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp (hunk 16 above)
Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp:110]: Yes, there is different behavior at n=0.

File: components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp (hunk 18 above)
Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp]: Update: @bogensch I just pushed a new commit that along with a fair amount of cleanup should add the tests we had talked about previously. Let me know if these weren't the ones you had in mind and feel free to add or request others.

Comment (bogensch) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp]: Thanks @pressel , yes the tests are definitely more what I imagined.  A question about how the input data is defined... Is it necessary that we do these tests with the 128 full model levels?  When I made modifications to the tests I was constructing a simple five layer model and explicitly defining the zt and zi grids (allowing for stretched grids to be tested) as well as the density varying with height as such (see screenshot).  This is more how I envisioned input data would be defined for most SHOC tests.  

Screen Shot 2020-06-30 at 11 08 40 AM

Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp]: @bogensch Yes, we can definitely specify the input this way.

Comment (pressel) [reply components/scream/src/physics/shoc/tests/shoc_grid_tests.cpp]: @bogensch, I implemented the test you mention above. Let me know if you want me to add any others.

**Comment by PeterCaldwell** [comment]
@pressel - I think the PR is all set. Is there a reason why you didn't include the AT:automerge label? Once you do that, I think the PR will get merged...
